sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/GlobalEndpointManager.java

```diff
@@ -340,4 +341,19 @@ public String getRegionName(URI locationEndpoint, OperationType operationType) {
     public ConnectionPolicy getConnectionPolicy() {
         return this.connectionPolicy;
     }
+
+    private List<String> getEffectivePreferredRegions() {
+
+        if (this.connectionPolicy.getPreferredRegions() != null && !this.connectionPolicy.getPreferredRegions().isEmpty()) {
+            return this.connectionPolicy.getPreferredRegions();
+        }
+
+        // when latestDatabaseAccount is initialized
```

Comment:
How i atomicity ensured here? WOuldn't you need to hold at least a read lock to read this.lastDatabaseAccount and locationCache atomically?

Reply:
So `latestDatabaseAccount` is `volatile` and `locationCache` is `final` and its contents (location-related properties are updated under a lock) so should be visible to other threads. So you're right there is no atomicity but only visibility - I can update to make `latestDatabaseAccount` and `locationCache` updates under a write lock and its access under a read lock.

Reply:
It will make database account refresh sequential but should be okay since such refreshes happen more frequently outside the hot path.

Reply:
So, what I did is plain access of `latestDatabaseAccount` - I leave it as is. Any reassignment I wrap it up with a write lock and when I have to access both `latestDatabaseAccount` and `locationCache` I do so under a read lock.

Reply:
Fixed in next iterations.